### PR TITLE
[Bugfix and Enhancement] fuse conv and bn into one optimized function using fx

### PR DESCRIPTION
**Pull request by youkaichao**
This PR tries to fix how conv and bn are fused in fx.

Prior to this PR, `torch.fx` fuses conv and bn **modules** into a new module, which has two drawbacks:
- It can be only used for inference.
- It does not work when conv is used in multiple places, leading to bugs like https://github.com/pytorch/pytorch/issues/104593 .

This PR takes an alternate approach: it tries to fuse conv and bn **computation nodes** into a new function. It is based on our recent paper "Tune-Mode ConvBN Blocks For Efficient Transfer Learning" (https://arxiv.org/abs/2305.11624): when bn is in `eval` mode (whether the whole network is in training or validation), consecutive conv and bn can be fused according to the associative law between convolution and affine transform: normalize (weight * feature) = (normalize weight) * feature, where * is the convolution operator.

This PR addresses the two drawbacks of `torch.fx` fuse implementation:
- It can be used for both training and inference. During inference, the bn and normalize process can be baked (constant-folded) into conv weights, which is the same as traditional conv-bn fusion. **During training, the bn normalizations are kept to stablize training, while being more efficient on the conv weight rather than the conv output feature**.
- Since this PR operates in the level of computation nodes, it does not modify the state dict of the module. It only captures the consecutive conv + bn computation pattern, and replaces it with a new function call. Therefore, it can deal with various corner cases mentioned in https://github.com/pytorch/pytorch/issues/104593 .

Here is a summary of computation graph change:

- Without any optimization, the computation graph is the one in `Eval` mode.
- Prior to this PR, `torch.fx` implements the `Deploy` mode, which changes the parameters and is incapable of dealing with multiple usage of one single conv module.
- This PR implements the `Tune` mode, which leaves the parameters untouched, and only changes the computation.

image

This PR has already been accepted in [mmcv](https://github.com/open-mmlab/mmcv) and [mmengine](https://github.com/open-mmlab/mmengine), and greatly benefits the training of object detectors in [mmdetection](https://github.com/open-mmlab/mmdetection) (with 20% memory cost reduction) without hurting the accuracy of models. I hope this PR can further benefit the pytorch community.

See the following tables for more results (this method is called Tune Mode).

image

image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by youkaichao** [comment]
@Chillee

**Comment by youkaichao** [comment]
@jansel I would also want to know if this optimization can get into the passes in pytorch 2.0. The `efficient_conv_bn_eval_graph_transform` function introduced in this PR can be used as one optimization pass to optimize the graph captured in torch dynamo. It seems pytorch 2.0 backends does not support composable optimization passes now. It only supports mutually exclusive compiler backends.

**Comment by jansel** [comment]
There are some similar fusions in TorchInductor here:
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/mkldnn_fusion.py
and
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/freezing_patterns.py

**Comment by vadimkantorov** [comment]
Maybe also related to inplace Conv+BatchNorm+LeakyReLU: https://github.com/pytorch/pytorch/issues/26288

**Comment by youkaichao** [comment]
> There are some similar fusions in TorchInductor here: [`main`/torch/_inductor/fx_passes/mkldnn_fusion.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/mkldnn_fusion.py?rgh-link-date=2023-08-01T18%3A21%3A48Z) and [`main`/torch/_inductor/fx_passes/freezing_patterns.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/freezing_patterns.py?rgh-link-date=2023-08-01T18%3A21%3A48Z)

Thanks for the pointers. They look related to this PR. Can you confirm if the exact fusion (general conv + bn in eval mode) is considered in TorchInductor now? As far as I can tell, it seems TorchInductor code you mentioned does not consider this fusion pattern. But I'm not sure.

**Comment by youkaichao** [comment]
@jansel Does the TorchInductor team have any plan to expose the passes API to users so that they can easily add new passes? Currently, users have to modify the pytorch codebase or submit a PR to add new passes. That's not very convenient.

I think a doable solution is to have a `CustomBackend` class, which leaves several interfaces like `graph_transform_forward`/`graph_transform_backward` etc. Then users can inherit or modify the class property to control the passes used:

```python
from typing import Any
import torch
from torch import nn
from torch import _dynamo
from torch._functorch.aot_autograd import aot_module_simplified
import functorch
from functorch.compile import make_boxed_func


class CustomBackend(object):
    pytorch_graph_passes = []
    forward_graph_passes = []
    backward_graph_passes = []

    def graph_processing_pytorch(self, gm, example_inputs):
        # graph transform (graph optimization) for the captured graph in pytorch graph
        for graph_pass in CustomBackend.pytorch_graph_passes:
            graph_pass(gm, example_inputs)

    def graph_processing_aot_forward(self, gm, example_inputs):
        # graph transform (graph optimization) for the captured graph in aot autograd forward graph
        for graph_pass in CustomBackend.forward_graph_passes:
            graph_pass(gm, example_inputs)

    def graph_processing_aot_backward(self, gm, example_inputs):
        # graph transform (graph optimization) for the captured graph in aot autograd backward graph
        for graph_pass in CustomBackend.backward_graph_passes:
            graph_pass(gm, example_inputs)

    def forward_compiler(self, gm, example_inputs):
        self.graph_processing_aot_forward(gm, example_inputs)
        return make_boxed_func(gm.forward)

    def backward_compiler(self, gm, example_inputs):
        self.graph_processing_aot_backward(gm, example_inputs)
        return make_boxed_func(gm.forward)


    def __call__(self, gm, example_inputs):
        self.graph_processing_pytorch(gm, example_inputs)
        return aot_module_simplified(
            gm, example_inputs,
            fw_compiler=self.forward_compiler,
            bw_compiler=self.backward_compiler
        )
```

**Comment by jansel** [comment]
If there is strong interest I'd be open to that, but am hesitant to add new APIs without clear users as the APIs can add a maintenance burden over time.

If the passes result in speedups and are sounds we do welcome pull requests to add passes directly to TorchInductor.

In your example above you should be able to just replace `make_boxed_func` with a call to `torch._inductor.compile()` (the direct compilation API that takes an FX graph) to get something pretty close to what you want.

**Comment by youkaichao** [comment]
@jansel Sorry for the late reply.

> If there is strong interest I'd be open to that, but am hesitant to add new APIs without clear users as the APIs can add a maintenance burden over time.

I think the passes are much like image transforms in torchvision, where torchvision have a library of transforms and users can select and compose the sequence of transforms that fit their usecases. Backend often refers to hardware related parts, and this single level of abstraction is clearly not enough. Optimization passes are the missing abstraction, and we will need it sooner or later.

> If the passes result in speedups and are sound we do welcome pull requests to add passes directly to TorchInductor.

The optimization pass in this PR has already been accepted in [mmcv](https://github.com/open-mmlab/mmcv) and [mmengine](https://github.com/open-mmlab/mmengine), and greatly benefits the training of object detectors in [mmdetection](https://github.com/open-mmlab/mmdetection) (with 20% memory cost reduction) without hurting the accuracy of models. The soundness of this PR is guaranteed by the corresponding paper [arxiv.org/abs/2305.11624](https://arxiv.org/abs/2305.11624).

As mentioned in https://github.com/pytorch/pytorch/issues/104593 , this PR fixes that bug with `torch.fx.experimental.optimization.fuse`. After this PR is merged, I can create another PR to torchinductor to add this pass. But perhaps @Chillee should review this PR first.

In addition, the code bug also leads to a misleading tutorial in https://pytorch.org/tutorials/intermediate/fx_conv_bn_fuser.html . I would like to improve that tutorial too, if @svekars thinks it is good to go.

**Comment by jansel** [comment]
cc @eellison

**Comment by eellison** [comment]
> Thanks for the pointers. They look related to this PR. Can you confirm if the exact fusion (general conv + bn in eval mode) is considered in TorchInductor now? As far as I can tell, it seems TorchInductor code you mentioned does not consider this fusion pattern. But I'm not sure.

Yes, I can confirm. You need to enable `TORCHINDUCTOR_FREEZING=1`. I also confirmed that https://github.com/pytorch/pytorch/issues/104593 works correctly in torchinductor. See this gist as an [example](https://gist.github.com/eellison/ce3440e5728849ce73ee96f683d17464). Note that we decompose batchnorm, and then fold add/sub/mul/div, which is the more general form of the optimization, and works with custom batchnorm classes like [this one](https://github.com/facebookresearch/maskrcnn-benchmark/blob/main/maskrcnn_benchmark/layers/batch_norm.py#L27-L31). 

I do think it would be nice to make these available eventually to other backends. 

You're welcome to talk with @svekars about https://pytorch.org/tutorials/intermediate/fx_conv_bn_fuser.html. I think that tutorial is more meant as illustrative of fx passes, then a definitive guide to conv-bn fusion. In that tutorial, it uses torch.jit.script, which has it's own conv-bn fusion pass anyway.

@Chillee can speak to maintaining the fx pass. 

I appreciate your interest in upstreaming and trying to fix existing issues. If you're looking for a way to contribute to torchinductor or other folding efforts, in Torchinductor we have yet to implement Linear-BN fusion and would accept a pr that enables that.

**Comment by youkaichao** [comment]
@eellison Thanks for the information! Actually, the gist example and the `TORCHINDUCTOR_FREEZING=1` setting are used for **inference without training**, and it cannot deal with the case that **weights need to be updated**. Please note that, the code in this PR, **supports training**.

**Comment by youkaichao** [comment]
@eellison I hope it is clear that conv-bn folding with `TORCHINDUCTOR_FREEZING=1` only works for inference, while this PR works for training :)

**Comment by youkaichao** [comment]
@eellison I would like to hear your further opinion to guide this PR :)

**Comment by youkaichao** [comment]
@jansel would you mind have a look at this PR when you have time?

**Comment by youkaichao** [comment]
@eellison @jansel @Chillee Would you mind take some time to have a look? I believe this PR is fundamentally different than existing optimization @eellison mentioned with `TORCHINDUCTOR_FREEZING=1`: this PR works for both **training** and inference, and `TORCHINDUCTOR_FREEZING=1` only works for inference.

The technique in this PR has helped to reduce memory footprint by 20%~40% in training object detectors, and incorporating it into pytorch can benefit further more users.

**Comment by jansel** [comment]
@eellison would be the best person to look at this.

Looks like there are lint/test failures

**Comment by eellison** [comment]
> Please note that, the code in this PR, supports training.

@youkaichao, since there is already inference covered, could you please demonstrate a performance win on training ?

It doesn't look like there is one: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_opinfo.py#L281

Also, as Jason mentioned, we already fuse batchnorm through torchinductor. 

it might be easier for you to have in a separate repro, and upstream when the benefits are clear. You can register graph passes at various places in the stack, without having to touch pytorch core. See, pre-grad, post-grad, joint-graph passes: https://github.com/pytorch/pytorch/tree/main/torch/_inductor/fx_passes.

**Comment by youkaichao** [comment]
@eellison The pre-grad, post-grad, joint-graph passes are great. I will look at it.

> could you please demonstrate a performance win on training

The performance win has been demonstrated in multiple cases. I add a test of the function, where you can arbitrarily vary the size of conv features to see the test. More examples can be found in the paper [arxiv.org/abs/2305.11624](https://arxiv.org/abs/2305.11624). I just copied the main result here, where the paper named it as "Tune Mode":

Here is how it helps for image classification:

image

Here is how it helps for object detection:

image

Here is the user feedback from OpenMMLab: https://github.com/open-mmlab/mmcv/pull/2807#issuecomment-1586824505 .

Please feel free to point out any additional usage cases you would like to see.

> It doesn't look like there is one: [main/test/inductor/test_torchinductor_opinfo.py#L281](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_opinfo.py?rgh-link-date=2023-08-24T19%3A51%3A24Z#L281)

This line points to a `rand_like` function. Can you elaborate on what do you mean by this line?

**Comment by youkaichao** [comment]
The `pre_grad_passes` are great. I can easily integrate this function into torchinductor. Now we can either use `torch.fx.experimental.efficient_conv_bn_eval.turn_on_efficient_conv_bn_eval` directly, or use this feature in torchinductor without bothering users. Both cases are tested in `test/fx/test_efficient_conv_bn_eval.py`.

**Comment by youkaichao** [comment]
@eellison I would like to hear your opinion first. This optimization changes the computation rather than the parameter, and some tests should be updated if we agree on the improvement.

In addition, I tried the example you mentioned before, `TORCHINDUCTOR_FREEZING` cannot deal with multiple uses of a single `conv` module, nor can it deal with many dynamic cases. Please check this [gist](https://gist.github.com/youkaichao/f2fab96fa121e197d40ad3726cff5c7f) for example. Note that the code in this PR works quite well in dealing with these corner cases.

The code in this PR can not only deal with many dynamic usage of conv+bn pairs, but it also works for training. The only one caveat is that for inference users, we have to add one additional constant folding step, which should be no problem because many runtimes have builtin support for constant folding (e.g. [torch.onnx.export](https://pytorch.org/docs/stable/onnx.html#torch.onnx.export) enables constant folding by default).

**Comment by youkaichao** [comment]
Closed as it has been moved to https://github.com/pytorch/pytorch/pull/108757 and been accepted there.
